File: packages/snow_draw_core/lib/draw/models/camera_state.dart
import 'package:meta/meta.dart';

import '../types/draw_point.dart';

@immutable
class CameraState {
  const CameraState({this.position = DrawPoint.zero, this.zoom = 1.0});

  final DrawPoint position;
  final double zoom;

  static const initial = CameraState();
  static const minZoom = 0.1;
  static const maxZoom = 30.0;

  static double clampZoom(double zoom) => zoom.clamp(minZoom, maxZoom);

  CameraState copyWith({DrawPoint? position, double? zoom}) =>
      CameraState(position: position ?? this.position, zoom: zoom ?? this.zoom);

  CameraState translated(double dx, double dy) => copyWith(
    position: DrawPoint(x: position.x + dx, y: position.y + dy),
  );

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CameraState && other.position == position && other.zoom == zoom;

  @override
  int get hashCode => Object.hash(position, zoom);

  @override
  String toString() => 'CameraState(position: $position, zoom: $zoom)';
}

File: packages/snow_draw_rust_engine/src/draw/models/camera_state.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::hash::{Hash, Hasher};

/// Camera transform state (pan + zoom) for the drawing viewport.
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
pub struct CameraState {
    /// Camera position in world-space coordinates.
    pub position: DrawPoint,

    /// Camera zoom scale.
    pub zoom: f64,
}

impl CameraState {
    /// Minimum supported zoom value.
    pub const MIN_ZOOM: f64 = 0.1;

    /// Maximum supported zoom value.
    pub const MAX_ZOOM: f64 = 30.0;

    /// Default camera state used on document/load reset.
    pub const INITIAL: Self = Self::new(DrawPoint::ZERO, 1.0);

    pub const fn new(position: DrawPoint, zoom: f64) -> Self {
        Self { position, zoom }
    }

    /// Clamps a zoom value into the valid camera range.
    pub fn clamp_zoom(zoom: f64) -> f64 {
        zoom.clamp(Self::MIN_ZOOM, Self::MAX_ZOOM)
    }

    /// Returns a copied state with selected fields replaced.
    pub fn copy_with(self, position: Option<DrawPoint>, zoom: Option<f64>) -> Self {
        Self {
            position: position.unwrap_or(self.position),
            zoom: zoom.unwrap_or(self.zoom),
        }
    }

    /// Returns a copied state translated by `(dx, dy)`.
    pub fn translated(self, dx: f64, dy: f64) -> Self {
        self.copy_with(
            Some(DrawPoint::new(self.position.x + dx, self.position.y + dy)),
            None,
        )
    }
}

impl Default for CameraState {
    fn default() -> Self {
        Self::INITIAL
    }
}

impl PartialEq for CameraState {
    fn eq(&self, other: &Self) -> bool {
        self.position == other.position && self.zoom == other.zoom
    }
}

impl Hash for CameraState {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.position.hash(state);
        self.zoom.to_bits().hash(state);
    }
}

impl fmt::Display for CameraState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "CameraState(position: {}, zoom: {})",
            self.position, self.zoom
        )
    }
}
